File: pat1036.cpp
/*
	思路：将男生和女生分别用两个vector存起来，分别排序即可
*/


#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
using namespace std;

struct Student
{
	string name;
	char gender;
	string ID;
	int grade;
};

bool comp(Student s1, Student s2) {
	return s1.grade < s2.grade;
}
int main() {

	int N;
	cin >> N;
	vector<Student> vf, vm;
	for (int i = 0; i < N; i++) {
		string name, ID;
		char gender;
		int grade;
		cin >> name >> gender >> ID >> grade;
		Student s;
		s.name = name;
		s.gender = gender;
		s.ID = ID;
		s.grade = grade;
		if (gender == 'F') {
			vf.push_back(s);
		}
		else {
			vm.push_back(s);
		}
	}
	bool isabsent = false;
	
	if (vf.size() == 0) {
		cout << "Absent" << endl;
		isabsent = true;
	}
	else {
		sort(vf.begin(), vf.end(), comp);
		cout << vf[vf.size() - 1].name << ' ' << vf[vf.size() - 1].ID << endl;
	}
	if (vm.size() == 0) {
		cout << "Absent" << endl;
		isabsent = true;
	}
	else {
		sort(vm.begin(), vm.end(), comp);
		cout << vm[0].name << ' ' << vm[0].ID << endl;
	}
	if (isabsent == true) {
		cout << "NA";
	}
	else {
		cout << vf[vf.size() - 1].grade - vm[0].grade;
	}
	return 0;
}

File: pat1036.py
'''
    思路：分两个列表装男生和女生，然后分别排序
'''

male_student = []
female_student = []
N = int(input())

for x in range(N):
    name,gender,ID,grade = input().split()
    student = {'name':name,'gender':gender,'ID':ID,'grade':int(grade)}
    if gender == 'M':
        male_student.append(student)
    else:
        female_student.append(student)

isabsent = False
if len(female_student) == 0:
    print('Absent')
    isabsent = True
else:
    highest_female = sorted(female_student,key=lambda x:x['grade'],reverse=True)[0]   #   获取成绩最高的女生
    print(highest_female['name'] + ' ' + highest_female['ID'])

if len(male_student) == 0:
    print('Absent')
    isabsent = True
else:
    lowest_male = sorted(male_student,key=lambda x:x['grade'])[0]   #   获取成绩最低的男生
    print(lowest_male['name'] + ' ' + lowest_male['ID'])

if isabsent == True:
    print('NA')
else:
    print(highest_female['grade'] - lowest_male['grade'])